src/compiler/expressionToTypeNode.ts

```diff
@@ -0,0 +1,504 @@
+        if (node.kind === SyntaxKind.GetAccessor) {
+            return createReturnFromSignature(node, context);
+        }
+        else {
+            context.tracker.reportInferenceFallback(node);
+            return false;
+        }
+    }
+
+    function typeFromTypeAssertion(expression: Expression, type: TypeNode, context: SyntacticTypeNodeBuilderContext, requiresAddingUndefined: boolean) {
+        if (isConstTypeReference(type)) {
+            return typeFromExpression(expression, context, /*isConstContext*/ true, requiresAddingUndefined);
+        }
+        if (requiresAddingUndefined && !canAddUndefined(type)) {
+            context.tracker.reportInferenceFallback(type);
+        }
+        return serializeExistingTypeAnnotation(type, context);
+    }
+    function typeFromExpression(node: Expression, context: SyntacticTypeNodeBuilderContext, isConstContext = false, requiresAddingUndefined = false, preserveLiterals = false): boolean | undefined {
+        switch (node.kind) {
+            case SyntaxKind.ParenthesizedExpression:
+                if (isJSDocTypeAssertion(node)) {
+                    return typeFromTypeAssertion(node.expression, getJSDocTypeAssertionType(node), context, requiresAddingUndefined);
+                }
+                return typeFromExpression((node as ParenthesizedExpression).expression, context, isConstContext, requiresAddingUndefined);
+            case SyntaxKind.Identifier:
+                if (context.isUndefinedIdentifier(node as Identifier)) {
+                    return true;
+                }
+                break;
+            case SyntaxKind.NullKeyword:
+                return true;
+            case SyntaxKind.ArrowFunction:
+            case SyntaxKind.FunctionExpression:
+                return typeFromFunctionLikeExpression(node as ArrowFunction | FunctionExpression, context);
+            case SyntaxKind.TypeAssertionExpression:
+            case SyntaxKind.AsExpression:
+                const asExpression = node as AsExpression | TypeAssertion;
+                return typeFromTypeAssertion(asExpression.expression, asExpression.type, context, requiresAddingUndefined);
+            case SyntaxKind.PrefixUnaryExpression:
+                const unaryExpression = node as PrefixUnaryExpression;
+                if (isPrimitiveLiteralValue(unaryExpression)) {
+                    if (unaryExpression.operand.kind === SyntaxKind.BigIntLiteral) {
+                        return typeFromPrimitiveLiteral();
+                    }
+                    if (unaryExpression.operand.kind === SyntaxKind.NumericLiteral) {
+                        return typeFromPrimitiveLiteral();
+                    }
+                }
+                break;
+            case SyntaxKind.NumericLiteral:
+                return typeFromPrimitiveLiteral();
+            case SyntaxKind.TemplateExpression:
+                if (!isConstContext && !preserveLiterals) {
+                    return true;

```

Comment:
...is this right? Eg, for
```ts
const a = "a";
export const b = `b${a}`
```
we emit
`export declare const b = "ba";` but for
```ts
let a = "a";
export const b = `b${a}`
```
we emit `export declare const b: string;` - you can't know if a template literal type should be emitted as a literal type (or initializer) or `string` without doing evaluation, and if that evaluation crosses file bounds... 🤷‍♂️ I'm pretty sure this has been a consideration, given the `evaluator` changes in this PR, it just doesn't seem to be reflected in the diagnostic.

Reply:
This is exactly the reason for the `!isConstContext && !preserveLiterals`. `preserveLiterals` is true for const vars and readonly properties. `isConstContext` is if we are in a const assertion. Both of these may result in preservation of template literal string types, so we don't even try since we can't determine if the type is preserved only based on the expression (non local type information is needed, as shown in your example)

So the only place where we can safely infer `string` is mutable variables and mutable property declarations.

```ts
const a = "a";
export const b = `b${a}` // Can't infer string based on local info
export let m = `b${a}` // Will be string 
export class Foo { 
    b = `b${a}` // Will be string 
    readonly ro = `b${a}` // Can't infer string based on local info
}
```

Template literals in const assertions are not really useful without non local information. I don't think doing something like `` `b${"a"}` `` or `` `b${"a" as "a" | "b"}` ``(which would only need local information to be evaluated) is common enough to be worth testing  for and that is the only  kind of template literal string we could get a template literal string type from.